netmisc.cc

```diff
@@ -327,73 +327,40 @@ static unsigned num_bits(long arg)
-		    /* We need this extra select to hide the signed
-		     * property from the padding above. It will be
-		     * removed automatically during code generation. */
-		  NetESelect *tmp = new NetESelect(base, 0 , min_wid);
-		  tmp->set_line(*base);
-		  tmp->cast_signed(true);
-                  base = tmp;
-	    }
-	      /* Normalize the expression. */
-	    base = make_sub_expr(offset+soff, base);
-      } else {
-	      /* Correct the offset if needed. */
-	    if (!is_up) offset += wid - 1;
-	      /* If the offset is zero then just return the base (index)
-	       * expression. */
-	    if ((soff-offset) == 0) return base;
-	      /* Calculate the space needed for the offset. */
-	    unsigned min_wid = num_bits(-offset);
-	    if (num_bits(soff) > min_wid)
-		  min_wid = num_bits(soff);
+      bool msb_lo = msb < lsb;
+      long offset = msb_lo ? soff + lsb : soff - lsb;
+
+	/* Correct the offset if needed. */
+      if (!msb_lo ^ is_up) offset -= wid - 1;
+
+	/* Calculate the space needed for the offset. */
+      unsigned off_wid = num_bits(offset);
+      unsigned off_max_wid = max(off_wid, num_bits(offset + wid - 1));
+	/* Save the width of the base expression. */
+      unsigned base_wid = base->expr_width();
+
+	/* Check for anything but zero offset bit selects. */
+      if (off_max_wid) {
 	      /* We need enough space for the larger of the offset or the
-	       * base expression. */
-	    if (min_wid < base->expr_width()) min_wid = base->expr_width();
-	      /* Now that we have the minimum needed width increase it by
-	       * one to make room for the normalization calculation. */
-	    min_wid += 2;
-	      /* Pad the base expression to the correct width. */
-	    base = pad_to_width(base, min_wid, *base);
-	      /* If the offset is greater than zero then we need to do
-	       * signed math to get the location value correct. */
-	    if (offset > 0 && ! base->has_sign()) {
-		    /* We need this extra select to hide the signed
-		     * property from the padding above. It will be
-		     * removed automatically during code generation. */
-		  NetESelect *tmp = new NetESelect(base, 0 , min_wid);
-		  tmp->set_line(*base);
-		  tmp->cast_signed(true);
-                  base = tmp;
-	    }
-	      /* Normalize the expression. */
-	    base = make_add_expr(base, soff-offset);
+	       * base expression, plus an extra bit for arithmetic overflow. */
+	    unsigned sum_wid = 1 + max(off_max_wid, base_wid);
+	    base = pad_to_width(base, sum_wid, *base);
+      }
+
+	/* There is no need to add a zero offset. */
+      if (!msb_lo && offset == 0) return base;
+
+	/* If the offset is negative or the result could otherwise be negative,
+	 * then we need to do signed math to get the location value correct. */
+      if ((offset < 0 || (msb_lo && off_wid <= base_wid)) && ! base->has_sign()) {
+	    unsigned signed_wid = 1 + base->expr_width();
+	    base = pad_to_width(base, signed_wid, *base);
+	    base->cast_signed(true);
```

Comment:
This is only safe if `off_max_wid` is non-zero and the earlier `pad_to_width()` has already zero-extended `base`. If not, the immediately preceding `pad_to_width()` will sign-extend `base`. As it happens, I think your extra expansion of `off_max_wid` guarantees it will never be zero, but if that is removed, you will need to restore the additional `NetESelect` here to separate the cast from the padding.

Reply:
It turns out that the `NetESelect` is necessary to generate the correct `%pad/u`, otherwise `cast_signed` will become part of the previous `pad_to_width`, and `%pad/s` will be generated.

It basically says so in the comments, I just didn't catch the meaning of it since I didn't know the code base and this somewhat counterintuitive behavior.